File: src/main/java/me/centralhardware/znatoki/telegram/statistic/telegram/callbackHandler/student/UserInfoCallback.java
package me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.student;

import lombok.RequiredArgsConstructor;
import me.centralhardware.znatoki.telegram.statistic.i18n.ErrorConstant;
import me.centralhardware.znatoki.telegram.statistic.i18n.MessageConstant;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.ServiceMapper;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.ServicesMapper;
import me.centralhardware.znatoki.telegram.statistic.service.ClientService;
import me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.CallbackHandler;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.CallbackQuery;
import org.telegram.telegrambots.meta.api.objects.User;

@Component
@RequiredArgsConstructor
public class UserInfoCallback extends CallbackHandler {

    private final ClientService clientService;

    private final ServiceMapper serviceMapper;
    private final ServicesMapper servicesMapper;

    @Override
    public void handle(CallbackQuery callbackQuery, User from, String data) {
        if (!telegramService.hasReadRight(from.getId())){
            sender.sendMessageFromResource(ErrorConstant.ACCESS_DENIED, from);
            return;
        }
        var clientOptional = clientService.findById(Integer.parseInt(data.replace("/user_info","")));
        clientOptional.ifPresentOrElse(
                client -> {
                    if (!client.getOrganizationId().equals(getTelegramUser(from).getOrganizationId())){
                        sender.sendText("Доступ запрещен", from);
                        return;
                    }

                    sender.sendMessageWithMarkdown(client.getInfo(serviceMapper.getServicesForCLient(client.getId()).stream().map(servicesMapper::getNameById).toList()), callbackQuery.getFrom());
                },
                () -> sender.sendMessageFromResource(MessageConstant.USER_NOT_FOUND, from)
        );
    }

    @Override
    public boolean isAcceptable(String data) {
        return data.startsWith("/user_info");
    }
}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/telegram/callbackHandler/student/UserInfoCallback.kt
package me.centralhardware.znatoki.telegram.statistic.telegram.callbackHandler.student

import dev.inmo.tgbotapi.extensions.api.send.sendMessage
import dev.inmo.tgbotapi.extensions.behaviour_builder.BehaviourContext
import dev.inmo.tgbotapi.extensions.behaviour_builder.triggers_handling.onDataCallbackQuery
import dev.inmo.tgbotapi.types.message.MarkdownParseMode
import me.centralhardware.znatoki.telegram.statistic.entity.StudentId
import me.centralhardware.znatoki.telegram.statistic.entity.getInfo
import me.centralhardware.znatoki.telegram.statistic.entity.toStudentId
import me.centralhardware.znatoki.telegram.statistic.mapper.StudentMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.LessonMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.SubjectMapper

fun BehaviourContext.userInfoCallback() = onDataCallbackQuery(Regex("user_info\\d+$")) {
    StudentMapper.findById(it.data.replace("user_info", "").toInt().toStudentId())?.let { client ->
        val info =
            client.getInfo(
                LessonMapper.getSubjectIdsForStudent(client.id!!)
                    .mapNotNull { SubjectMapper.getNameById(it) }
                    .toList()
            )
        sendMessage(it.from, info, parseMode = MarkdownParseMode)
    } ?: sendMessage(it.from, "Пользователь не найден")
}
